Transitioning from global street variable to redux street

mkong0216 opened the pull request:
The application no longer uses or references the global `street` variable! (Resolves #799) 

Instead, new action creators/reducers have been added for updating the Redux `street` state and are used throughout the application. When the entire `street` object needs to be updated, there is a function called `updateStreetData` that calls several action creators in order to update the entire `street` object. 

A fix has also been made for fixing the undo/redo action of updating building heights. (Resolves #800)

To do in separate PR:
- Remove the store subscriber and call `saveStreetToServerIfNecessary` within action creators
- Attempt to find a fix for the janky quality of resizing segments. It appears that when dragging the resize block/button, the block moves faster than the actual resizing of the segment.

louh commented:
This is a great effort, @mkong0216. I'm really looking forward to getting this merged.

Some thoughts.

1. With the change to `StreetWidth` (renaming it to `StreetMetaWidth`) could you update this branch and incorporate the latest on `master` branch?
2. For functions that make a side-effect call to `store.getState().street` (e.g. in `assets/scripts/app/page_title.js`) we might consider getting that information further back in the call stack and passing `street` in as a argument to those function instead. For instance what if `updatePageTitle()` was `updatePageTitle(street)`? My hope is that by doing this, these functions are easier to test and work with, and the modules that they're in don't need to rely on importing the Redux store.

In regards to the jank related to resizing segments, it seems like the resize action causes very expensive React updates to street segments, and it may be related to `recalculateWidth()` in `assets/scripts/streets/`. This is just a hunch, but there's a `store.dispatch()` inside a `for` loop that goes through each segment; each dispatch would cause something else to re-render, and the jank seems to pile up so that a street with only 1 segment is less janky than a street with 10 segments. So one potential optimization is to see if that `dispatch()` happen only once regardless of the number of segments. This may not solve all of the jankiness, but if we can reduce the cost of one expensive function that would very helpful.

mkong0216 commented:
@louh No problem, updated branch.

Also, for the second point, what do you mean by "side-effect calls to `store.getState().street`"? Do you mean to find the files that only imports `store` to check its state? So for files that import `store` in order to `dispatch` actions, would it be okay to have functions that call `store.getState().street`? Sorry, not quite sure what other functions besides `updatePageTitle` to change. 

For example, for `prettifyWidth` I noticed that the file `assets/scripts/util/width_units.js` that the function is in only calls store to get the `street.units` when the `units` parameter is equal to `undefined` so I just passed in `store.getState().street` to the `prettyWidth` function calls in `view.js` and `resizing.js`, and removed `store` from `assets/scripts/util/width_units.js` is that alright?

mkong0216 commented:
**UPDATE:** So I think I fixed the jank related to resizing segments! Instead of dispatching `updateSegmentWarnings` within the `for` loop, I used `updateSegments` outside the `for` loop. I definitely see a difference. Thanks @louh

louh commented:
That definitely made a huge difference!!

louh commented:
OK, I think this is pretty great!

I did a measurement to compare before and after and the old resize listener took about 6ms to do stuff, and the new code takes about 60ms. This means it's about 10x slower, but since the biggest bottlenecks during the segment resize are removed, I'm still comfortable accepting this code so far. Now that the global `street` variable is gone, we'll be able to port the street segments to React, and by rewriting stuff I'm confident we'll be able to optimize this more until we're back to about 60fps. All this can happen subsequently after merging this PR.

One last thing (a regression). Steps to reproduce:

- Requirements: user is logged in; user has a street that is "Unnamed St", and a street that has a name.
- Open the street that has a name.
- Open the gallery by clicking "My streets" on the menu.
- Click on to the street that is Unnamed, which switches the street displayed.

Expected result: the street that the user switches to should change to show "Unnamed St."
Actual result: the street that the user switches to retains the street name from the previous street.

mkong0216 commented:
@lou Ahh, yea that's because the `street` object is no longer completely replaced when switching between gallery streets so if the previous street's name was user updated, then the switched gallery street wasn't able to override it if it was using the `DEFAULT_NAME`. 

The previous commits fixes the regression by modifying the `saveStreetName` action creator and reducer to have a `system` parameter which defaults to false and if the renaming of the street is caused by the system then it will update the street name despite the previous `userUpdated` value.
